src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,258 @@
+                    var request = new PutObjectRequest
+                    {
+                        BucketName = options.BucketName,
+                        ContentType = artifactSettings.ContentType,
+                        Key = artifactSettings.Name,
+                        InputStream = stream,
+                        AutoCloseStream = false,
+                    };
+                    await client.PutObjectAsync(request, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    using TransferUtility transferUtility = new(client);
+                    await transferUtility.UploadAsync(stream, options.BucketName, artifactSettings.Name, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        private string GetResourceId(IAmazonS3 client, S3StorageEgressProviderOptions options, EgressArtifactSettings artifactSettings)
+        {
+            if (!options.GeneratePresSignedUrl)
+                return $"BucketName={options.BucketName}, Key={artifactSettings.Name}";
+
+            DateTime expires = DateTime.UtcNow.AddMinutes(options.PreSignedUrlExpiryInMinutes);
+            string resourceId = client.GetPreSignedURL(new GetPreSignedUrlRequest { BucketName = options.BucketName, Key = artifactSettings.Name, Expires = expires });
+            Logger?.EgressProviderSavedStream(EgressProviderTypes.S3Storage, resourceId);
+            return resourceId;
+        }
+
+        private static async Task<IAmazonS3> CreateClientAsync(S3StorageEgressProviderOptions options, CancellationToken cancellationToken)
+        {
+            AWSCredentials awsCredentials = null;
+            AmazonS3Config configuration = new();
+            // use the specified access key and the secrets taken from configuration or a local file
+            if (!string.IsNullOrEmpty(options.AccessKeyId) && (!string.IsNullOrEmpty(options.SecretAccessKey) || !string.IsNullOrEmpty(options.SecretsAccessKeyFile)))
+            {
+                string secretAccessKeyId;
+                if (!string.IsNullOrEmpty(options.SecretsAccessKeyFile) && File.Exists(options.SecretsAccessKeyFile))
+                    secretAccessKeyId = await WrapException(async () => (await File.ReadAllTextAsync(options.SecretsAccessKeyFile, cancellationToken)).Trim());
+                else
+                    secretAccessKeyId = options.SecretAccessKey;
+                awsCredentials = new BasicAWSCredentials(options.AccessKeyId, secretAccessKeyId);
+
+                configuration.ForcePathStyle = options.ForcePathStyle;
+                if (!string.IsNullOrEmpty(options.Endpoint))
```

Comment:
Is it necessary to check if null or empty rather than just assigning to the properties for both `Endpoint` and `RegionName`?